### PR TITLE
Improve API docs

### PR DESCRIPTION
**Pull request by chamil321**
## Checklist
- [ ] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests

### CODE REVIEW

File: http-ballerina/http_service_endpoint.bal [hunk 1]
```diff
@@ -62,19 +59,19 @@ public class Listener {
 
     # Attaches a service to the listener.
     #
-    # + s - The service that needs to be attached
+    # + httpService - The service that needs to be attached
     # + name - Name of the service
     # + return - An `error` an error occurred during the service attachment process or else nil
```
Comment (MadhukaHarith92) [inline http-ballerina/http_service_endpoint.bal:64]: Let's add the Oxford comma before “and” or “or” in lists.

File: http-ballerina/Module.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+facilitates two types of network entry points as `Client` and `Listener`.
+
+### Client
+
+The `Client` is used to connect to and interact with HTTP endpoints. They support connection pooling and can be 
+configured to have a maximum number of active connections that can be made with the remote endpoint. The `Client` 
+activates connection eviction after a given idle period and also supports follow-redirects so that the users do not 
+have to manually handle 3xx HTTP status codes.
+
+The `Client` handles resilience in multiple ways such as load balancing, circuit breaking, endpoint timeouts, and a
```
Comment (shafreenAnfar) [inline http-ballerina/Module.md:13]: I think we can have a high level section as Resiliency

File: http-ballerina/Module.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+When a failure occurs in the remote service, the client connections might wait for some time before a timeout occurs. 
+Awaiting requests consume resources in the system. Circuit Breakers are used to trip after a certain number of failed 
+requests to the remote service. Once a circuit breaker trips, it does not allow the client to send requests to the 
+remote service for a period of time.
+
+The Ballerina circuit breaker supports tripping on HTTP error status codes and I/O errors. Failure thresholds can be 
+configured based on a sliding window (e.g., 5 failures within 10 seconds). `Client` also supports a retry 
+mechanism that allows a client to resend failed requests periodically for a given number of times.
+
+The `Client` supports Server Name Indication (SNI), Certificate Revocation List (CRL), Online Certificate Status
```
Comment (shafreenAnfar) [inline http-ballerina/Module.md:27]: Then we need a section for security and explain the rest as well in addition to SSL level stuff.

Comment (shafreenAnfar) [reply http-ballerina/Module.md:27]: Also I think talk bit about Caching

File: http-ballerina/Module.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+mechanism that allows a client to resend failed requests periodically for a given number of times.
+
+The `Client` supports Server Name Indication (SNI), Certificate Revocation List (CRL), Online Certificate Status 
+Protocol (OCSP), and OCSP Stapling for SSL/TLS connections. They also support HTTP2, keep-alive, chunking, HTTP 
+caching, data compression/decompression, and authentication/authorization.
+
+A `Client` can be defined using the URL of the remote service that the client needs to connect with, as shown below:
+
+```ballerina
+http:Client|http:ClientError clientEndpoint = new("https://my-simple-backend.com");
```
Comment (shafreenAnfar) [inline http-ballerina/Module.md:34]: Lets write it with `check` statement.

File: http-ballerina/Module.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+http:Response|http:ClientError response = clientEndpoint->get("/get?id=123");
+```
+The payload can be retrieved as return value from the remote function as follows:
+
+```ballerina
+// Retrieve payload as json.
+json payload = check clientEndpoint->post("/backend/Json", "foo", targetType = json);
+```
+
+### Listener
```
Comment (shafreenAnfar) [inline http-ballerina/Module.md:49]: I think we can explain the type oriented approach we took here.

File: http-ballerina/Package.md [hunk 6]
```diff
@@ -1,73 +1,20 @@
 ## Package Overview
 
-This package provides an implementation for connecting and interacting with HTTP and HTTP2 endpoints. The package facilitates two types of network entry points as ‘Client’ and ‘Listener’.
+The HTTP library is one of the standard library packages of the<a target="_blank" href="https://ballerina.io/"> Ballerina</a> language.
 
-### Client
+It provides an implementation for connecting and interacting with HTTP, HTTP2, and WebSocket endpoints over the network.
 
-The `Client` is used to connect to and interact with HTTP endpoints. They support connection pooling and can be configured to have a maximum number of active connections that can be made with the remote endpoint. The `Client` activates connection eviction after a given idle period and also supports follow-redirects so that the users do not have to manually handle 3xx HTTP status codes.
+For more information go to [The HTTP Package](https://ballerina.io/learn/api-docs/ballerina/http/index.html).
```
Comment (daneshk) [inline http-ballerina/Package.md:7]: Let's remove API Doc and example links from package.md

File: README.md [hunk 7]
```diff
@@ -94,7 +94,7 @@ All contributors are encouraged to read the [Ballerina Code of Conduct](https://
 
 ## Useful Links
 
-* Discuss about code changes of the Ballerina project in [ballerina-dev@googlegroups.com](mailto:ballerina-dev@googlegroups.com).
+* Discuss the code changes of the Ballerina project in [ballerina-dev@googlegroups.com](mailto:ballerina-dev@googlegroups.com).
```
Comment (MadhukaHarith92) [inline README.md:97]:
```suggestion
* Discuss code changes of the Ballerina project in [ballerina-dev@googlegroups.com](mailto:ballerina-dev@googlegroups.com).
```

File: http-ballerina/Module.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+The `Listener` is the underneath server connector that binds the given IP/Port to the network, and it's behaviour can 
+be changed using the `http:ListenerConfiguration`. In HTTP, the `http:Service` typed services can be attached to 
+the `listener`. The service type precisely describes the syntax for both service and resource.
+
+A `Service` represents a collection of network-accessible entry points and can be exposed via a `Listener` endpoint. 
+A resource represents one such entry point and can have its own path, HTTP methods, body format, `consumes` and 
+`produces` content types, CORS headers, etc. In resources, Http method and resource path are mandatory parameters and
```
Comment (praneesha) [inline http-ballerina/Module.md:65]:
```suggestion
`produces` content types, CORS headers, etc. In resources, the HTTP method and resource path are mandatory parameters and
```

File: http-ballerina/Module.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+String literal and path parameters can be stated as path. The resource function accepts `http:Caller`, `http:Request`,
```
Comment (praneesha) [inline http-ballerina/Module.md:66]:
```suggestion
the String literal and path parameters can be stated as the path. The resource function accepts the `http:Caller`, `http:Request`, 
```

File: http-ballerina/Module.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+The `Listener` is the underneath server connector that binds the given IP/Port to the network, and it's behaviour can 
+be changed using the `http:ListenerConfiguration`. In HTTP, the `http:Service` typed services can be attached to 
+the `listener`. The service type precisely describes the syntax for both service and resource.
+
+A `Service` represents a collection of network-accessible entry points and can be exposed via a `Listener` endpoint. 
+A resource represents one such entry point and can have its own path, HTTP methods, body format, `consumes` and 
+`produces` content types, CORS headers, etc. In resources, Http method and resource path are mandatory parameters and
+String literal and path parameters can be stated as path. The resource function accepts `http:Caller`, `http:Request`, 
+`http:Headers`, Query parameters, Header parameters, and Payload parameters as arguments, but they are optional.
```
Comment (praneesha) [inline http-ballerina/Module.md:67]:
```suggestion
`http:Headers`, query parameters, header parameters, and payload parameters as arguments. However, they are optional.
```

File: http-ballerina/Module.md [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+       return "Hello, World! I’m " + name + ". " + message;
+   }
+}
+```
+
+#### Security
+
+`Listener` endpoints can be exposed via SSL. They support Mutual SSL, Hostname Verification, and Application Layer 
+Protocol Negotiation (ALPN) for HTTP2. `Listener` endpoints also support Certificate Revocation List (CRL), Online 
+Certificate Status Protocol (OCSP), OCSP Stapling.
```
Comment (praneesha) [inline http-ballerina/Module.md:95]:
```suggestion
Certificate Status Protocol (OCSP), and OCSP Stapling.
```

File: http-ballerina/Module.md [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+Also, The `listener` can be configured to authenticate and authorize the inbound requests. Ballerina has 
+built-in support for basic authentication, JWT authentication, and OAuth2 authentication.
+
+In addition to that, protocol wise it supports both HTTP/1.1 and HTTP2 and feature wise connection keep-alive, content
```
Comment (praneesha) [inline http-ballerina/Module.md:99]:
```suggestion
In addition to that, supports both the HTTP/1.1 and HTTP2 protocols and connection keep-alive, content 
```

File: http-ballerina/Module.md [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+#### Security
+
+`Listener` endpoints can be exposed via SSL. They support Mutual SSL, Hostname Verification, and Application Layer 
+Protocol Negotiation (ALPN) for HTTP2. `Listener` endpoints also support Certificate Revocation List (CRL), Online 
+Certificate Status Protocol (OCSP), OCSP Stapling.
+Also, The `listener` can be configured to authenticate and authorize the inbound requests. Ballerina has 
+built-in support for basic authentication, JWT authentication, and OAuth2 authentication.
+
+In addition to that, protocol wise it supports both HTTP/1.1 and HTTP2 and feature wise connection keep-alive, content 
+chunking, HTTP caching, data compression/decompression, payload binding, and authorization can be highlighted.
```
Comment (praneesha) [inline http-ballerina/Module.md:100]:
```suggestion
chunking, HTTP caching, data compression/decompression, payload binding, and authorization can be highlighted as the features of a `Service`.
```

File: http-ballerina/Package.md [hunk 14]
```diff
@@ -1,73 +1,15 @@
 ## Package Overview
 
-This package provides an implementation for connecting and interacting with HTTP and HTTP2 endpoints. The package facilitates two types of network entry points as ‘Client’ and ‘Listener’.
+The HTTP library is one of the standard library packages of the<a target="_blank" href="https://ballerina.io/"> Ballerina</a> language.
```
Comment (praneesha) [inline http-ballerina/Package.md:3]:
```suggestion
The `http` library is one of the standard library packages of the<a target="_blank" href="https://ballerina.io/"> Ballerina</a> language.
```

File: http-ballerina/http2_push_promise.bal [hunk 15]
```diff
@@ -46,7 +46,7 @@ public class PushPromise {
     # If there are more than one header value for the specified header name, the first value is returned.
     #
     # + headerName - The header name
-    # + return - The header value, or null if there is no such header
+    # + return - The header value, or nil if there is no such header
```
Comment (praneesha) [inline http-ballerina/http2_push_promise.bal:49]:
```suggestion
    # + return - The header value or `()` if there is no such header
```

File: http-ballerina/http_client.bal [hunk 16]
```diff
@@ -16,16 +16,15 @@
...
 import ballerina/jballerina.java;
 
-# Provides the HTTP actions for interacting with an HTTP server. Apart from the standard HTTP methods,
-# `HttpClient.forward()` and `HttpClient.execute()` functions are provided. More complex and specific endpoint types
-# can be created by wrapping this generic HTTP actions implementation.
+# Lies inside every type of client in the chain holding the native client connector. More complex and specific
+# endpoint types are created by wrapping this generic HTTP actions implementation internally.
 #
 # + url - The URL of the remote HTTP endpoint
 # + config - The configurations associated with the HttpClient
```
Comment (praneesha) [inline http-ballerina/http_client.bal:23]:
```suggestion
# + config - The configurations associated with the `HttpClient`
```

File: http-ballerina/http_client.bal [hunk 17]
```diff
@@ -16,16 +16,15 @@
...
 public client class HttpClient {
 
-    public ClientConfiguration config = {};
     public string url;
+    public ClientConfiguration config = {};
 
     # Gets invoked to initialize the native `client`. During initialization, the configurations are provided through the
```
Comment (praneesha) [inline http-ballerina/http_client.bal:29]:
```suggestion
    # Gets invoked to initialize the native `client`. During the initialization, the configurations are provided through the
```

File: http-ballerina/http_client_endpoint.bal [hunk 18]
```diff
@@ -19,10 +19,6 @@ import ballerina/jballerina.java;
 import ballerina/observe;
 import ballerina/time;
 
-////////////////////////////////
-///// HTTP Client Endpoint /////
-////////////////////////////////
-
 # The HTTP client provides the capability for initiating contact with a remote HTTP service. The API it
 # provides includes functions for the standard HTTP methods, forwarding a received request and sending requests
```
Comment (praneesha) [inline http-ballerina/http_client_endpoint.bal:23]:
```suggestion
# provides includes the functions for the standard HTTP methods forwarding a received request and sending requests
```

File: http-ballerina/http_client_endpoint.bal [hunk 19]
```diff
@@ -365,21 +361,6 @@ public type TargetService record {|
 # Following fields are inherited from the other configuration records in addition to the Client specific
```
Comment (praneesha) [inline http-ballerina/http_client_endpoint.bal:361]:
```suggestion
# The following fields are inherited from the other configuration records in addition to the `Client`-specific
```

File: http-ballerina/http_client_endpoint.bal [hunk 20]
```diff
@@ -365,21 +361,6 @@ public type TargetService record {|
...
-# | forwarded - Copied from CommonClientConfiguration       |
-# | followRedirects - Copied from CommonClientConfiguration |
-# | poolConfig - Copied from CommonClientConfiguration      |
-# | cache - Copied from CommonClientConfiguration           |
-# | compression - Copied from CommonClientConfiguration     |
-# | auth - Copied from CommonClientConfiguration            |
-# | circuitBreaker - Copied from CommonClientConfiguration  |
-# | retryConfig - Copied from CommonClientConfiguration     |
-# | cookieConfig - Copied from CommonClientConfiguration    |
 # + secureSocket - SSL/TLS related options
```
Comment (praneesha) [inline http-ballerina/http_client_endpoint.bal:364]:
```suggestion
# + secureSocket - SSL/TLS-related options
```

File: http-ballerina/http_headers.bal [hunk 21]
```diff
@@ -30,7 +30,7 @@ public class Headers {
     # Checks whether the requested header key exists in the header map.
     #
     # + headerName - The header name
-    # + return - Returns true if the specified header key exists
+    # + return - True if the specified header key exists
```
Comment (praneesha) [inline http-ballerina/http_headers.bal:33]:
```suggestion
    # + return - `true` if the specified header key exists
```

File: http-ballerina/http_request.bal [hunk 22]
```diff
@@ -131,7 +131,7 @@ public class Request {
     # Checks whether the requested header key exists in the header map.
     #
     # + headerName - The header name
-    # + return - Returns true if the specified header key exists
+    # + return - True if the specified header key exists
```
Comment (praneesha) [inline http-ballerina/http_request.bal:134]:
```suggestion
    # + return - `true` if the specified header key exists
```

File: http-ballerina/http_request.bal [hunk 23]
```diff
@@ -193,7 +193,7 @@ public class Request {
 
     # Checks whether the client expects a `100-continue` response.
     #
-    # + return - Returns true if the client expects a `100-continue` response
+    # + return - True if the client expects a `100-continue` response
```
Comment (praneesha) [inline http-ballerina/http_request.bal:196]:
```suggestion
    # + return - `true` if the client expects a `100-continue` response
```

File: http-ballerina/http_response.bal [hunk 24]
```diff
@@ -184,7 +184,7 @@ public class Response {
 
     # Gets the type of the payload of the response (i.e: the `content-type` header value).
```
Comment (praneesha) [inline http-ballerina/http_response.bal:185]:
```suggestion
    # Gets the type of the payload of the response (i.e., the `content-type` header value).
```

File: http-ballerina/http_service_endpoint.bal [hunk 25]
```diff
@@ -62,19 +59,19 @@ public class Listener {
 
     # Attaches a service to the listener.
     #
-    # + s - The service that needs to be attached
+    # + httpService - The service that needs to be attached
     # + name - Name of the service
-    # + return - An `error` an error occurred during the service attachment process or else nil
-    public isolated function attach(Service s, string[]|string? name = ()) returns error? {
-        return externRegister(self, s, name);
+    # + return - An `error` if an error occurred during the service attachment process or else nil
```
Comment (praneesha) [inline http-ballerina/http_service_endpoint.bal:64]:
```suggestion
    # + return - An `error` if an error occurred during the service attachment process or else `()`
```

File: http-ballerina/http_service_endpoint.bal [hunk 26]
```diff
@@ -62,19 +59,19 @@ public class Listener {
...
+    public isolated function attach(Service httpService, string[]|string? name = ()) returns error? {
+        return externRegister(self, httpService, name);
     }
 
     # Detaches a Http service from the listener.
```
Comment (praneesha) [inline http-ballerina/http_service_endpoint.bal:69]:
```suggestion
    # Detaches an HTTP service from the listener.
```

File: http-ballerina/http_status_code_types.bal [hunk 27]
```diff
@@ -14,6 +14,7 @@
 // specific language governing permissions and limitations
 // under the License.
 
+// Remove the union once https://github.com/ballerina-platform/ballerina-lang/issues/30490 is fixed
```
Comment (praneesha) [inline http-ballerina/http_status_code_types.bal:17]:
```suggestion
// Remove the union once https://github.com/ballerina-platform/ballerina-lang/issues/30490 is fixed.
```

File: http-ballerina/http_status_code_types.bal [hunk 28]
```diff
@@ -14,6 +14,7 @@
...
 # Defines the possible Status code response record types.
```
Comment (praneesha) [inline http-ballerina/http_status_code_types.bal:18]:
```suggestion
# Defines the possible status code response record types.
```

File: http-ballerina/resiliency_failover_client.bal [hunk 29]
```diff
@@ -513,23 +513,6 @@ isolated function populateErrorsFromLastResponse (Response inResponse, ClientErr
 # Following fields are inherited from the other configuration records in addition to the failover client specific
```
Comment (praneesha) [inline http-ballerina/resiliency_failover_client.bal:513]:
```suggestion
# The following fields are inherited from the other configuration records in addition to the failover client-specific
```

File: http-ballerina/resiliency_load_balance_client.bal [hunk 30]
```diff
@@ -372,23 +372,6 @@ isolated function populateGenericLoadBalanceActionError(LoadBalanceActionErrorDa
 # The configurations related to the load balance client endpoint. Following fields are inherited from the other
```
Comment (praneesha) [inline http-ballerina/resiliency_load_balance_client.bal:372]:
```suggestion
# The configurations related to the load balancing client endpoint. The following fields are inherited from the other
```

File: http-ballerina/resiliency_load_balance_client.bal [hunk 31]
```diff
@@ -372,23 +372,6 @@ isolated function populateGenericLoadBalanceActionError(LoadBalanceActionErrorDa
...
 # configuration records in addition to the load balance client specific configs.
```
Comment (praneesha) [inline http-ballerina/resiliency_load_balance_client.bal:373]:
```suggestion
# configuration records in addition to the load balancing client specific configs.
```

File: http-ballerina/resiliency_load_balance_client.bal [hunk 32]
```diff
@@ -372,23 +372,6 @@ isolated function populateGenericLoadBalanceActionError(LoadBalanceActionErrorDa
...
-# | cache - Copied from CommonClientConfiguration           |
-# | compression - Copied from CommonClientConfiguration     |
-# | auth - Copied from CommonClientConfiguration            |
-# | circuitBreaker - Copied from CommonClientConfiguration  |
-# | retryConfig - Copied from CommonClientConfiguration     |
-# | cookieConfig - Copied from CommonClientConfiguration    |
-# | responseLimits - Copied from CommonClientConfiguration  |
-#
 # + targets - The upstream HTTP endpoints among which the incoming HTTP traffic load should be distributed
 # + lbRule - LoadBalancing rule
```
Comment (praneesha) [inline http-ballerina/resiliency_load_balance_client.bal:376]:
```suggestion
# + lbRule - The `LoadBalancing` rule
```

File: http-ballerina/resiliency_load_balance_client.bal [hunk 33]
```diff
@@ -372,23 +372,6 @@ isolated function populateGenericLoadBalanceActionError(LoadBalanceActionErrorDa
...
 # + failover - Configuration for load balancer whether to fail over in case of a failure
```
Comment (praneesha) [inline http-ballerina/resiliency_load_balance_client.bal:377]:
```suggestion
# + failover - Configuration for the load balancer whether to fail over a failure
```

**Review by MadhukaHarith92 — APPROVED** [review]
LGTM

**Pull request merged**